### PR TITLE
KAFKA-12648: Pt. 1 - Add NamedTopology to protocol and state directory structure

### PR DESCRIPTION
**Pull request by ableegoldman**
Pt. 1: [#10609](https://github.com/apache/kafka/pull/10609)
Pt. 2: [#10683](https://github.com/apache/kafka/pull/10683)
Pt. 3: [#10788](https://github.com/apache/kafka/pull/10788)

This PR includes adding the NamedTopology to the Subscription/AssignmentInfo, and to the StateDirectory so it can place NamedTopology tasks within the hierarchical structure with task directories under the NamedTopology parent dir.

### CODE REVIEW

**Comment by ableegoldman** [comment]
cc @wcarlson5

File: streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java]: Are we going to to just make all subtopologies named ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: I am assuming that this will be done in a later PR

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:378]: I think I saw a comment about updating these with named topologies later. Is there are reason you are waiting?

**Review by wcarlson5 — APPROVED** [review]
This makes sense. I don't see anything that worries me and won't be cleaned/finished in the follow ups, so I think we are good to merge!

File: streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java]: I have considered a bit whether to do that, and just make the "normal" (unnamed) case like a specific case of the named topology where the name is just the app id or empty string or something, but I think that ultimately only complicates things even more, so I decided against it. For example we'd need to check whether the named topology was equal to `app.id`/`""` before placing task dirs into named subdirectories, inserting the name into internal topics, etc. It's easier to just check whether it exists, that way it's pretty obvious if you miss something, vs a subtle bug that eg changed the internal topic names by one extra `-` 

But once we circle back on making this a public API I think we can clean things up a lot, at which time it may make more sense to consolidate things in that way

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: Yep, that's covered by [Pt. 2](https://github.com/apache/kafka/pull/10683) -- the PR is available for review, but it's on top of this PR so it's probably going to be difficult to review until this one is merged

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:378]: That comment was more about usages in integration/other unit tests, where we may want to tie in some NamedTopologyies once the basic feature is fully implemented. For the StateDirectoryTest everything should already be implemented so we should have the NamedTopology logic in that class covered by tests in this PR (actually I need to add a few more, I think)

File: streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/TaskId.java]: Once we make this a public API would make them all named but I am okay with waiting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:558]: Is it possible that we can have named topology state dirs and unamed (original) state dirs co-exist here?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Could we just remove empty named topology dirs along the way instead of doing that in a second pass at the end? 

EDIT: nvm, after some thoughts I feel it is more complicated than easier.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Should we move this into the try/catch IOException block as well (ditto below)?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java]: Could you remind me why we want to include the partition id in the new version as well?

**Review by guozhangwang — COMMENTED** [review]
Made a quick pass over the non-testing code, left some clarification questions.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:558]: No, that should not be allowed.  We have checks to verify this in a few places where it matters, but it's an assumption we can make here. I'm not sure if your question was from an implementation point of view or a semantic one, but I can further clarify or justify why it should not be allowed if you want

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Yeah, we could but I don't feel there is anything to gain, and the code would only be less easy to follow imo

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java]: Ah, yes. I tried to explain this with a comment on the `SubscriptionInfoData.json` schema but I'll call it out again in the `SubscriptionInfo.java` class. Previously we encoded the offset sums as a nested "map" of <topicGroupId, <partition, offsetSum>>, where the "map" is really an array and the array struct does not itself allow for struct types. It's just a gap in the API that no one has cared or had time to close, not a fundamental principle. Anyways this meant we had a TopicGroupId and a PartitionToOffsetSum struct, where in turn the PartitionToOffsetSum was composed of the partition and offset sum base types.

I guess this was reasonable enough when there were only 3 base fields, but if we wanted to maintain this nested array structure it would mean adding more and more nested structs each time we added a field. I felt this would get to be too complicated and annoying to deal with so I flattened the OffsetSum struct out to just include each base field directly

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:20]: Just wanted to lay out my test plan somewhere, so it doesn't seem like I'm merging all this code with no intention to ever test it. Once the final pieces are in (should be by Pt. 2) these are the things I think are important to touch on with integration tests. Leave a comment if you have any more suggestions or feedback 🙂

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java]: I've added a few tests for the named topology stuff but I definitely want to add more and just haven't yet had time. Since I'm still on-call and therefore unlikely to have time until next week, if you both are able to do a quick pass and don't have any further feedback on the PR as-is, it may make sense to just merge this PR tomorrow (Friday) and do a quick followup PR for the tests next week.

That way  I can rebase the next PR  ([Pt. 2](https://github.com/apache/kafka/pull/10683)) and you all can actually begin reviewing that. cc @guozhangwang @wcarlson5

**Comment by ableegoldman** [comment]
Rebased after the TaskId changes in KIP-470, and responded to all comments. Not much has changed since the last review, just cleaning up here and there. ~It's pretty much done except for StateDirectoryTest, which I can always do in a quick followup PR to unblock other downstream work with this~ edit: tests are done, this PR is fully ready for review and merge

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:390]: Just want to call this out since it's a change in behavior unrelated to this PR -- actually just something we could/should have cleaned up after removing the lock/file based locking. Previously we couldn't ever delete empty task dirs by cleaner thread (due to that Windows bug), now we can, so we should not exclude empty dirs here

**Review by wcarlson5 — APPROVED** [review]
I don't have any problems with the recent changes so I am still +1

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java]: Thanks for the explanation!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:558]: I was asking more for a semantic one -- as long as this is not expected then I'm happy for this piece as is :)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:656]: Could we add a test case to verify that in case both named topology dir and non-named topology dir co-exist, we would at certain step check against and throw?

**Review by guozhangwang — APPROVED** [review]
Just another minor comment about covering the not-allowed co-existence of both named and non-named topologies, otherwise LGTM.

**Comment by ableegoldman** [comment]
Thanks @guozhangwang ! I'm going to go ahead and merge this one now since it's been out there for so long, I will follow up on that test case in the Pt. 2 PR

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:656]: Ack, will add this test in the next PR

**Pull request merged**
